cache.go

```diff
@@ -0,0 +1,839 @@
+			w.Write(value)
+			respCC := parseCacheControlHeaders(result.Header)
+			if respCC.neverCache() || len(value) < int(clnt.minSize) {
+				if cacheErr == nil {
+					go func() {
+						clnt.api.RemoveObject(clnt.bucket, key)
+					}()
+				}
+				return
+			}
+			rs := result.Header.Get(xhttp.ContentRange)
+			if rs != "" {
+				// Avoid caching range GET's for now.
+				if cacheErr == nil {
+					go func() {
+						clnt.api.RemoveObject(clnt.bucket, key)
+					}()
+				}
+				return
+			}
+
+			go func() {
+				opts := getPutOpts(result.Header)
+				_, err := clnt.api.PutObject(clnt.bucket, key, bytes.NewReader(value), int64(len(value)), opts)
+				if err != nil {
+					clnt.up = false
+					logger.LogIf(context.Background(), err, "Failed to cache object")
+				}
+			}()
+			return
+		}
+	}
+}
+
+type cacheConfig struct {
+	endpoint  string
+	useTLS    bool
+	accessKey string
+	secretKey string
+	bucket    string
+	minSize   uint64
+	duration  time.Duration
+}
+
+func newCacheConfig() *cacheConfig {
+	cURL := os.Getenv(EnvCacheEndpoint)
+	if cURL == "" {
+		return nil
+	}
+	accessKey := os.Getenv(EnvCacheAccessKey)
+	secretKey := os.Getenv(EnvCacheSecretKey)
+	bucket := os.Getenv(EnvCacheBucket)
+	if accessKey == "" || secretKey == "" || bucket == "" {
+		console.Fatalln(fmt.Errorf("One or more of AccessKey:%s SecretKey: %s Bucket:%s missing", accessKey, secretKey, bucket), "Missing cache configuration")
+	}
+	minSizeStr := os.Getenv(EnvCacheMinSize)
+	var minSize uint64
+	var err error
+	if minSizeStr != "" {
+		minSize, err = humanize.ParseBytes(minSizeStr)
+		if err != nil {
+			console.Fatalln(fmt.Errorf("Unable to parse SIDEKICK_CACHE_MIN_SIZE %s should be in human readable units such as 64MB", minSizeStr))
+		}
+	}
+	durationStr := os.Getenv(EnvCacheHealthCheckDuration)
+	duration := defaultMinioHealthCheckDuration
+	if durationStr != "" {
+		duration, err = strconv.Atoi(durationStr)
+		if err != nil {
+			console.Fatalln(fmt.Errorf("Unable to parse SIDEKICK_CACHE_HEALTH_DURATION %s should be an integer", durationStr))
+		}
+	}
+	return &cacheConfig{endpoint: cURL,
+		accessKey: accessKey,
+		secretKey: secretKey,
+		bucket:    bucket,
+		minSize:   minSize,
+		duration:  time.Duration(duration) * time.Second,
+	}
+}
+func newCacheClient(ctx *cli.Context, cfg *cacheConfig) *S3CacheClient {
+	if cfg == nil {
+		return nil
+	}
+	creds := credentials.NewStaticV4(cfg.accessKey, cfg.secretKey, "")
+	var e error
+	s3Clnt := &S3CacheClient{}
+	options := minio.Options{
+		Creds:        creds,
+		Secure:       cfg.useTLS,
+		Region:       "",
+		BucketLookup: 0,
+	}
+	u, err := url.Parse(cfg.endpoint)
+	if err != nil {
+		return nil
+	}
+	api, e := minio.NewWithOptions(u.Host, &options)
+	if e != nil {
+		return nil
+	}
+	tr := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+		}).DialContext,
+		MaxIdleConns:          1024,
```

Comment:
Too big values @poornas this, as we have seen, doesn't help at all better to keep this value as 256